File: Phan_C/baitap47/baitap47.jl
using Printf

function doc_du_lieu(ten_tep::String)
    content = read(ten_tep, String)
    tokens = split(content)
    if isempty(tokens)
        return [], 0
    end
    n = parse(Int, tokens[1])
    d = parse(Int, tokens[2])
    D = []
    idx = 3
    for _ in 1:n
        features = [parse(Float64, tokens[i]) for i in idx:(idx + d - 1)]
        idx += d
        label = String(tokens[idx])
        idx += 1
        push!(D, (features, label))
    end
    return D, d
end

khoang_cach(u, v) = sqrt(sum((u .- v) .^ 2))

function lang_gieng_gan_nhat(q, D)
    min_dist = Inf
    best_label = ""
    best_idx = -1

    for (i, (features, label)) in enumerate(D)
        dist = khoang_cach(q, features)
        if dist < min_dist
            min_dist = dist
            best_label = label
            best_idx = i
        end
    end
    return best_label, best_idx, min_dist
end

function main()
    D, d = doc_du_lieu("hoa30.txt")
    if isempty(D) return end

    content = read(stdin, String)
    tokens = split(content)
    if isempty(tokens) return end

    idx = 1
    while idx <= length(tokens)
        q = [parse(Float64, tokens[i]) for i in idx:(idx + d - 1)]
        idx += d
        nhan, chi_so, dist = lang_gieng_gan_nhat(q, D)
        @printf("%s - mau %d, khoang cach %.4f\n", nhan, chi_so, dist)
        flush(stdout)
    end
end

main()

File: Phan_C/baitap47/baitap47.py
import sys
import math

def doc_du_lieu(ten_tep: str):
    D = []
    with open(ten_tep, 'r', encoding='utf-8') as f:
        lines = f.read().split()
        if not lines:
            return [], 0
        n = int(lines[0])
        d = int(lines[1])
        idx = 2
        for _ in range(n):
            features = [float(x) for x in lines[idx:idx + d]]
            idx += d
            label = lines[idx]
            idx += 1
            D.append((features, label))
    return D, d

def khoang_cach(u, v):
    return math.sqrt(sum((a - b) ** 2 for a, b in zip(u, v)))

def lang_gieng_gan_nhat(q, D):
    min_dist = float('inf')
    best_label = ""
    best_idx = -1

    for i, (features, label) in enumerate(D, start=1):
        dist = khoang_cach(q, features)
        if dist < min_dist:
            min_dist = dist
            best_label = label
            best_idx = i

    return best_label, best_idx, min_dist

def main():
    D, d = doc_du_lieu("hoa30.txt")
    if not D:
        return

    content = sys.stdin.read().split()
    if not content:
        return

    idx = 0
    while idx < len(content):
        q = [float(x) for x in content[idx:idx + d]]
        idx += d
        nhan, chi_so, dist = lang_gieng_gan_nhat(q, D)
        print(f"{nhan} - mau {chi_so}, khoang cách {dist:.4f}", flush=True)

if __name__ == "__main__":
    main()